src/main/scala/org/bitcoins/core/util/Base58.scala

```diff
@@ -0,0 +1,81 @@
+package org.bitcoins.core.util
+
+import scala.annotation.tailrec
+import scala.util.{Try, Failure, Success}
+
+/**
+  * Created by chris on 5/16/16.
+  * source of values: https://en.bitcoin.it/wiki/Base58Check_encoding
+  */
+trait Base58 extends BitcoinSLogger {
+
+  val base58Characters = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
+  val base58Pairs = base58Characters.zipWithIndex.toMap
+
+  /**
+    * Verifies a given base58 string against its checksum (last 4 decoded bytes)
+    * @param input base58 string
+    * @return decoded bytes excluding the checksum
+    */
+  def decodeCheck(input: String) : Try[Seq[Byte]] = {
+
+    val decoded : Seq[Byte] = decode(input)
+    if (decoded.length < 4) Failure(new IllegalArgumentException("Invalid input"))
+    else {
+      val splitSeqs = decoded.splitAt(decoded.length - 4)
+      val data : Seq[Byte] = splitSeqs._1
+      val checksum : Seq[Byte] = splitSeqs._2
+      val actualChecksum : Seq[Byte] = CryptoUtil.doubleSHA256(data).slice(0, 4)
+      if (checksum == actualChecksum)
+      Success(data)
+      else Failure(new IllegalArgumentException("checksums don't validate"))
+    }
+  }
+
+
+  /**
+    * Takes in sequence of bytes and returns base58 bitcoin string
+    * @param bytes sequence of bytes to be encoded into base58
+    * @return base58 String
+    */
+
+  def encode(bytes : Seq[Byte]) : String = {
+    val head : Byte = bytes.head
+    if (bytes.isEmpty) ""
+    else {
+      val big : BigInt = BigInt(1, bytes.toArray)
+      @tailrec
+      def loop(current : BigInt, str : String) : String = current match {
```

Comment:
It is usually best to define inner functions as the first thing inside of a parent function i.e. 

def encode(bytes : Seq[Byte] = {
  @tailrec 
  def loop(current : BigInt, str : String ) : String = {
     ...
  }
 //do other stuff that calls loop down here
}